### DISCUSSION TITLE
Adding e.g. Localization after the project was created?

### DISCUSSION THREAD

**Discussion by DevTKSS** [unoplatform/uno#18667] [Q&A] [answered]
Hey there,
I did init my project without Localization and now decided I would need it. Is there any way to add this later and when yes, how?
Thanks in advance!

**Comment by morning4coffe-dev** [comment, marked as answer]
Hey @DevTKSS,

Thank you for your question. Do you use Uno.Extensions in your project? We have a [guide](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Localization/LocalizationOverview.html) on how to add Uno.Extensions.Localization to your project without the hassle of implementing your own solution. Let me know if this helps in your case.

**Reply by DevTKSS** [reply]
Had to check, but yes, its in the global using file named, but only the dependency Injection, Hosting, Logging and Options. Tryed to add it in the csproj in features and added the folders in "strings"-Folder described in the tutorials for string localization, but didn't came much future before getting stuck in a kind of related problem. To catch the context of my localization usage is a RequiredInput Textbox control, which would be genious to have as a template (thinking templated Control = Figmas Templates but not sure if its better as a custom control?) and beside the idea to directly insert it in a Library Project, which I also never did, but followed the guides to that in the docs, which did make it not really much clearer or better said... a kind of complete guide would be nice, like they did on counter app or similar. The description of what is a template control or what is  a custom control was not finalized clear, at least for me. 

So writing that templated Textbox I followed the Idea to directly implement some Localization to the text of them.

Additional I had the "problem" to install a binding similar to the c# markup easy to do binding, but seems that is not possible in xaml what is quite frustrating that I have to add a whole class definition using IValueConverter to just do this from c#:

```
                    new NavigationBar().Content(() => vm.Title),
                    new StackPanel()
                        .Grid(row: 1)
                        .HorizontalAlignment(HorizontalAlignment.Center)
                        .VerticalAlignment(VerticalAlignment.Center)
                        .Spacing(16)
                        .Children(
                            new TextBox()
                                .Name(out var UserNameBox)
                                .Text(x => x.Binding(() => vm.Name).Mode(BindingMode.TwoWay))
                                .PlaceholderText("Enter your name:"),
                            new Button()
                                .Content("Go to Second Page")
                                .AutomationProperties(automationId: "SecondPageButton")
                                .Command(() => vm.GoToSecond),
                            new Button()
                                .Name(out var GoRightButton)
                                .Content("Let's go right")
                                .AutomationProperties(automationId: "GoRightButton")
                                .Command(() => vm.ShowDialogAsync)
                                .IsEnabled(x => x.Source(UserNameBox)
                                                 .Binding(() => !string.IsNullOrEmpty(UserNameBox.Text))) //<- desired Binding here
                                ),
```
just in xaml because the actual c# markup app with defining tasks in MainModel seems to have a bug.

my xaml attempt to use the Localizer [Localizer from How to localize text ressources](https://platform.uno/docs/articles/guides/localization.html) was this:
```
<ResourceDictionary
xmlns="http://schemas.microsoft.com/winfx/2006/xaml/presentation"
xmlns:x="http://schemas.microsoft.com/winfx/2006/xaml"
xmlns:utu="using:Uno.Toolkit.UI"
xmlns:local="using:tks_controllibrary">
  <!--<local:EmptyStringToBooleanConverter x:Key="EmptyStringToBooleanConverter"/>-->
  <Style TargetType="local:TextBoxWithRequiredInput">
    <Setter Property="Template">
      <Setter.Value>
        <ControlTemplate TargetType="local:TextBoxWithRequiredInput">
          <utu:AutoLayout Orientation="Vertical"
                          Spacing="5">
            <TextBox x:Name="InputTextBox"
                     Text="{TemplateBinding Text}"
                     PlaceholderText="{TemplateBinding PlaceholderText}"
                     HorizontalAlignment="Stretch"/>
            <TextBlock x:Name="ErrorTextBlock"
                       x:Uid="TextBoxWithRequiredInput_ErrorMessage"
                       Visibility="{Binding ElementName=InputTextBox}"/> //no Idea how to create the binding correct in the xaml, seems like I missed the help topic in the documenation to this.
          </utu:AutoLayout>
        </ControlTemplate>
      </Setter.Value>
    </Setter>
  </Style>
</ResourceDictionary>
```
with this xaml.cs behind:
```
using System;
using System.Collections.Generic;
using System.Globalization;
using System.Linq;
using System.Runtime.InteropServices.WindowsRuntime;
using System.Security.Cryptography.X509Certificates;
using Microsoft.Extensions.Localization;
using Microsoft.UI.Xaml;
using Microsoft.UI.Xaml.Controls;
using Microsoft.UI.Xaml.Data;
using Microsoft.UI.Xaml.Documents;
using Microsoft.UI.Xaml.Input;
using Microsoft.UI.Xaml.Media;
using Uno.Extensions.Localization;

// To learn about building custom controls see 
// https://learn.microsoft.com/windows/apps/winui/winui3/xaml-templated-controls-csharp-winui-3

namespace tks_controllibrary;
public sealed partial class TextBoxWithRequiredInput : TextBox
{
    public TextBoxWithRequiredInput()
    {
        this.DefaultStyleKey = typeof(TextBoxWithRequiredInput);
        this.Loaded += (s, e) =>
        {
            InputTextBox.Text = ResourceLoader.GetForViewIndependentUse().GetString("TextBoxWithRequiredInput_ErrorMessage")
        };
    }
}
```
and to get the result from string.IsNullOrEmpty of the Named InputTextbox I tryed to use the here descriped [IValueConverter](https://platform.uno/docs/articles/external/uno.extensions/doc/Learn/Markup/Converters.html) which is not showing any definition of the actual class they generated for that "public static readonly IValueConverter" in all of that code on this side.
```
public static class Converters
{
    public static IValueConverter StringToBool = new StringToBoolConverter();

    private object StringToBoolConverter()
    {
        public object Convert(object value, Type targetType, object parameter, CultureInfo culture)
        {
            if (value is string str)
            {
                return !string.IsNullOrEmpty(str);
            };
            return false;
        };
        public object ConvertBack(object value, Type targetType, object parameter, CultureInfo culture)
        {
            throw new NotImplementedException();
        };
    }
}
```
other attempt to get this result without IValueConverter:
```
        <Button Content="Start a cool simple Game!"
                AutomationProperties.AutomationId="SecondPageButton"
                Click="{x:Bind StartButton_Click}"
                Command="{Binding GoToStart}"
                HorizontalAlignment="Stretch"
                IsEnabled="{x:Bind tbName.Text.Length}"/>
```
but if I try to add a operation mark after Length like >0 the compiler defintly dont wants it there.
At least lets add just what I experienced while researching in docs:
It seems that related topics (Localization also) are spread across different parts of the documentation and not everytime matching. e.g. the Localization you linked now I did not find but the one in the tutorials which is telling something different than the one in extensions, like the one in extensions is... kind of on a old state? Seems like it's not even using the Localization or its just what I am thinking after overviewing, while the one in tutorials seems kind of beginner friendlyer and easyer to wrap the mind around. But I could not tell if one of them is false or which one should be taken in choice for such attempt. Maybe the one in tutorial, but if you try out the code I tryed in the Loaded Event you will see, that compiler defintly dont likes it althought the Localization feature is noted in my csproj file and listed in the using on the page start. It do not recognize the RessourceLoader althought I closly checked the writing. seems like there should be an instance because its giving me only the option for Equals or String which normally means exacly that and would change if a class is instanced. Also the x:Name of the TextBox in my Templated Control is not found from codebehind what it should be.

Hope you might could help
